File: R_Programming/ProgrammingAssignment3/rankall.R
rankall <- function (outcome, num='best') {
    list_outcome  <- list("heart failure" = 17, "heart attack" = 11, "pneumonia" = 23)
    loc_outcome <- list_outcome[[outcome]]
    if (is.null(loc_outcome)) stop('invalide outcome')
    
    outcome_file  <- 'Hospital/outcome-of-care-measures.csv'
    hospital_file <- 'Hospital/hospital-data.csv'
    
    stats <- read.csv(outcome_file)
    columns <- colnames(stats)
    
    # convert the loc_outcome to numeric and split out the NA entries
    stats[, loc_outcome] <- suppressWarnings(as.numeric(levels(stats[, loc_outcome]))[stats[, loc_outcome]])
    sub_stats <- subset(stats, !is.na(stats[, loc_outcome]), select=c(columns[2],columns[7],columns[loc_outcome]))
    
    rank <- function(df) {
        if (is.character(num)) {
            list_rank <- list('best'=1, 'worst'=nrow(df))
            num <- list_rank[[num]]
        }
        # if there is no qualified entry or num exceeds the maximal number of entries 
        # return NA
        # otherwise rank the data by mortality rate and hospital name and 
        # select the num-th entry
        # only the hospital.name is returned
        if (nrow(df) == 0 || nrow(df) < num) NA
        else {
            as.character(df[ order(df[3], df[1]), ][num, c(1)])
        }
    }
    ret <- lapply(split(sub_stats, sub_stats$State), rank)

    # return a 2-column data frame, column_1:hospital column_2:state
    data.frame(hospital=as.character(ret), state=names(ret))
}

File: R_Programming/ProgrammingAssignment3/rankall.py
import pandas as pd
import numpy as np
from os import path, listdir

def rankall(outcome, num='best'):
    """
    rank hospital for each state based on the mortality of the outcome
    """
    outcome_file = 'outcome-of-care-measures.csv'
    if not path.exists(outcome_file):
        outcome_file = path.join('Hospital', outcome_file)

    stats = pd.read_csv(outcome_file)
    columns = stats.columns.values
    loc_name = 1
    loc_state = 6

    dict_outcome = {'heart attack': 10, 'heart failure': 16, 'pneumonia': 22}
    try:
        loc_outcome = dict_outcome[outcome]
    except KeyError:
        raise Exception('invalid outcome')

    # convert the column to numerics
    stats.iloc[:, loc_outcome] = stats.iloc[:, loc_outcome].convert_objects(convert_numeric=True)
    # drop the NAs
    sub_stats = stats.dropna(axis=0, subset=[columns[loc_outcome]]).iloc[:, [loc_name,loc_state,loc_outcome]]

    # handler for ranking the hospitals
    def ranker(df, num):
        if not isinstance(num, int):
            dict_num = {'best':0, 'worst':len(df) - 1}
            num = dict_num[num]
        else:
            num -= 1
        # if num exceeds the number of entries for the state
        # return None    
        if num > len(df) - 1:
            ret = pd.DataFrame({'Hospital.Name':None, \
                                'State':df.State.unique()[0]}, \
                               index=np.arange(1))
        else:
            df.set_index(np.arange(len(df)), inplace=True)
            ret = pd.DataFrame({'Hospital.Name':df.ix[num][0], \
                                'State':df.ix[num][1]}, \
                               index=np.arange(1))
        return ret
        
    sub_stats.sort([columns[loc_outcome], columns[loc_name]], ascending=True, inplace=True)

    # wrap up the ranker with a lambda function
    by_state = sub_stats.groupby('State').apply(lambda x:ranker(x, num))

    # clean the multi-index of by_state
    return by_state.set_index(by_state.index.get_level_values('State'))
    
    

